ntoskrnl/mm/ARM3/mdlsup.c

```diff
@@ -1622,29 +1622,297 @@ MmAdvanceMdl(IN PMDL Mdl,
+    //
+    ASSERT(MappingAddress);
+    PointerPte = MiAddressToPte(MappingAddress) - 2;
+    ASSERT(!PointerPte[0].u.Hard.Valid &&
+           !PointerPte[1].u.Hard.Valid);
+
+    //
+    // Verify that the pool tag matches
+    //
+    TempPte.u.Long = PoolTag;
+    TempPte.u.Hard.Valid = 0;
+    if (PointerPte[1].u.Long != TempPte.u.Long)
+    {
+        KeBugCheckEx(SYSTEM_PTE_MISUSE,
+                     0x104, /* Trying to map an address it does not own */
+                     (ULONG_PTR)MappingAddress,
+                     PoolTag,
+                     PointerPte[1].u.Long);
+    }
+
+    //
+    // We must have a size, and our helper PTEs must be invalid
+    //
+    if (PointerPte[0].u.List.NextEntry < 3)
+    {
+        KeBugCheckEx(SYSTEM_PTE_MISUSE,
+                     0x105, /* Trying to map an invalid address */
+                     (ULONG_PTR)MappingAddress,
+                     PoolTag,
+                     (ULONG_PTR)_ReturnAddress());
+    }
+
+    //
+    // If the mapping isn't big enough, fail
+    //
+    if (PointerPte[0].u.List.NextEntry - 2 < PageCount)
+    {
+        DPRINT1("Reserved mapping too small. Need %Iu pages, have %Iu\n",
+                        PageCount,
+                        PointerPte[0].u.List.NextEntry - 2);
+        return NULL;
+    }
+
+    //
+    // Skip our two helper PTEs
+    //
+    PointerPte += 2;
+
+    //
+    // Get the template
+    //
+    TempPte = ValidKernelPte;
+    switch (CacheAttribute)
+    {
+        case MiNonCached:
+
+            //
+            // Disable caching
+            //
+            MI_PAGE_DISABLE_CACHE(&TempPte);
+            MI_PAGE_WRITE_THROUGH(&TempPte);
+            break;
+
+        case MiWriteCombined:
+
+            //
+            // Enable write combining
+            //
+            MI_PAGE_DISABLE_CACHE(&TempPte);
+            MI_PAGE_WRITE_COMBINED(&TempPte);
+            break;
+
+        default:
+            //
+            // Nothing to do
+            //
+            break;
+    }
+
+    //
+    // Loop all PTEs
+    //
+    do
+    {
+        //
+        // We're done here
+        //
+        if (*MdlPages == LIST_HEAD) break;
+
+        //
+        // Write the PTE
+        //
+        TempPte.u.Hard.PageFrameNumber = *MdlPages;
+        MI_WRITE_VALID_PTE(PointerPte++, TempPte);
+    } while (++MdlPages < LastPage);
+
+    //
+    // Mark it as mapped
+    //
+    ASSERT((Mdl->MdlFlags & MDL_MAPPED_TO_SYSTEM_VA) == 0);
+    Mdl->MappedSystemVa = MappingAddress;
+    Mdl->MdlFlags |= MDL_MAPPED_TO_SYSTEM_VA;
+
+    //
+    // Check if it was partial
+    //
+    if (Mdl->MdlFlags & MDL_PARTIAL)
+    {
+        //
+        // Write the appropriate flag here too
+        //
+        Mdl->MdlFlags |= MDL_PARTIAL_HAS_BEEN_MAPPED;
+    }
+
+    //
+    // Return the mapped address
+    //
+    return (PVOID)((ULONG_PTR)MappingAddress + Mdl->ByteOffset);
 }
 
 /*
- * @unimplemented
+ * @implemented
```

Comment:
It would be great too if you will add some Doxygen documentation for this and other newly implemented functions, describing the purpose of the function and its parameters & return value, basing on example here: https://reactos.org/wiki/Documentation_Guidelines. (optionally, but nice to have) You can read the information on MSDN as well, but you need to use your own wording. Thank you!

Reply:
I do not feel competent to add the Doxygen documentation. I am not familiar with them at all, so I would like to leave this to a more knowledgeable person in another PR. Thanks.